File: services/orchestrator/internal/pool/claim.lua
-- claim.lua — claim MỘT pod ấm từ warm-pool (phase-1 1.A-2).
--
-- ATOMIC NGHĨA LÀ GÌ Ở ĐÂY — đọc kỹ, đừng suy ra nhiều hơn:
-- Redis chạy script đơn luồng nên KHÔNG lệnh nào của client khác chen vào giữa
-- (isolation). Nhưng Redis Lua KHÔNG CÓ ROLLBACK: nếu một `redis.call` lỗi ở
-- giữa, mọi ghi TRƯỚC đó được giữ nguyên và replicate. "Chạy trọn hoặc chưa
-- chạy" là SAI nếu chỉ trông vào cơ chế của Redis.
--
-- Vì vậy script này TỰ hoàn tác: mọi ghi đi qua `w()` (dùng `redis.pcall`), và
-- khi có lỗi thì `undo()` đưa pod về lại `pool:free` rồi trả error_reply. Đó là
-- lý do phần thân dài hơn một chuỗi lệnh thẳng — chuỗi lệnh thẳng để lại pod
-- nằm trong `pool:claimed` mà không session nào trỏ tới, đúng loại trạng thái
-- mà reaper phải đoán. (Đo được: EXPIRE lỗi vì ttl quá lớn → session ghi xong
-- mà KHÔNG có TTL, giữ pod vĩnh viễn, trong khi caller nhận error và retry.)
--
-- KEYS[1] = pool:free        (LIST, FIFO — D6)
-- KEYS[2] = pool:claimed     (LIST, index cho reaper dò pod mồ côi)
-- KEYS[3] = session:{id}     (HASH — SSOT trạng thái session)
-- KEYS[4] = session:{id}:pod (STRING — con trỏ SỐNG LÂU HƠN hash, cho reaper)
-- KEYS[5] = pool:quarantine  (LIST, pod hỏng/trùng — KHÔNG tự quay lại pool)
--
-- ARGV[1] = sessionID      ARGV[5] = nowUnix
-- ARGV[2] = userID         ARGV[6] = expiresAtUnix
-- ARGV[3] = namespace      ARGV[7] = ttlSeconds
-- ARGV[4] = tier           ARGV[8] = podKeyPrefix (từ rediskeys — SSOT, không hardcode ở đây)
--                          ARGV[9] = podPointerTTL (> ttlSeconds; xem KEYS[4])
--
-- Trả: podName khi claim được; false/nil khi pool RỖNG (sentinel cho cold-path,
-- KHÔNG phải lỗi); error_reply khi có sự cố thật (đã hoàn tác xong).
--
-- GIỚI HẠN CLUSTER (D10): `pod:{name}` dựng động từ giá trị LMOVE trả về nên
-- không khai báo được trong KEYS. Trên Redis Cluster script này sẽ CROSSSLOT.
-- P1 chạy Redis đơn — chấp nhận có chủ ý; lên Cluster thì hash-tag `{dlp}`.

local podPrefix = ARGV[8]

-- Gọi hai lần với cùng sessionID là LỖI LẬP TRÌNH, không phải idempotency:
-- dedupe theo idempotency_key sống ở tầng trên (`idem:{userId}:{key}`). Không
-- chặn ở đây thì lần hai ghi đè hash và pod của lần một rò VĨNH VIỄN — nó vẫn
-- nằm trong pool:claimed nhưng `session:{id}` TỒN TẠI nên heuristic mồ côi của
-- reaper (pod trong claimed mà không có session) không bao giờ thấy.
if redis.call('EXISTS', KEYS[3]) == 1 then
  return redis.error_reply('claim: session da ton tai, khong claim de len')
end

-- ── Chọn pod: rút tới khi gặp một pod THẬT SỰ đang free ──────────────────────
-- Không tin `pool:free` một cách mù quáng. Một tên pod lọt vào list hai lần
-- (replenish retry sau timeout, reaper trả pod hai lần, hai instance cùng
-- replenish) sẽ được claim hai lần, và HAI sinh viên exec vào CÙNG một pod —
-- cả hai đều qua authz vì hash của mỗi người ghi đúng userId của người đó.
-- Redis LIST không chống trùng, nên chỗ chống phải là đây.
local MAX_SCAN = 16
local pod = nil
for _ = 1, MAX_SCAN do
  local candidate = redis.call('LMOVE', KEYS[1], KEYS[2], 'LEFT', 'RIGHT')
  if not candidate then
    return nil -- pool rỗng thật → cold path
  end

  -- Trong Lua chuỗi rỗng là TRUTHY, nên `if not candidate` KHÔNG bắt được ''.
  -- Bỏ qua thì sinh ra key rác tên đúng `pod:` và một session trỏ vào hư không.
  if candidate == '' then
    redis.call('LREM', KEYS[2], 1, '')
  else
    -- pcall chứ không call: nếu `pod:{name}` sai kiểu (string thay vì hash) thì
    -- `HGET` trả WRONGTYPE và một `redis.call` sẽ ABORT script NGAY — sau khi
    -- LMOVE đã chạy, tức là để lại đúng cái pod mồ côi mà toàn bộ phần hoàn
    -- tác bên dưới tồn tại để ngăn. Sai kiểu ⇒ pod hỏng ⇒ cách ly, như mọi
    -- pod không-free khác.
    local state = redis.pcall('HGET', podPrefix .. candidate, 'state')
    if state == 'free' then
      pod = candidate
      break
    end
    -- Không free (đã claimed / hash biến mất / tên trùng): cách ly.
    -- KHÔNG đẩy lại pool:free — làm thế là vòng lặp vô tận trên cùng pod hỏng.
    redis.call('LREM', KEYS[2], 1, candidate)
    redis.call('RPUSH', KEYS[5], candidate)
  end
end

if not pod then
  -- Quét hết MAX_SCAN mà toàn pod hỏng: coi như pool rỗng (cold path vẫn phục
  -- vụ được sinh viên, còn lỗi thì không). `pool:quarantine` dài ra là tín hiệu
  -- cho ops — B9 đếm nó.
  return nil
end

-- ── Ghi trạng thái, có hoàn tác ──────────────────────────────────────────────
local failed = nil

local function w(...)
  if failed then return end
  local r = redis.pcall(...)
  if type(r) == 'table' and r.err then failed = r.err end
end

local function undo()
  -- Thứ tự ngược lại. Dùng redis.call (không pcall): nếu chính bước hoàn tác
  -- lỗi thì im lặng còn tệ hơn — để nó nổi lên.
  redis.call('DEL', KEYS[3], KEYS[4])
  redis.call('HSET', podPrefix .. pod, 'state', 'free')
  redis.call('LREM', KEYS[2], 1, pod)
  -- LPUSH (không RPUSH): trả pod về ĐẦU TRÁI, đúng chỗ nó vừa bị rút ra, nên
  -- thứ tự FIFO của pool không bị xáo sau một lần abort.
  redis.call('LPUSH', KEYS[1], pod)
end

-- userId/tier ghi CẢ vào hash pod, không chỉ hash session.
--
-- KHÔNG phải nhân bản thừa: khi `session:{id}` hết hạn, hash đó BIẾN MẤT, và
-- reaper cần biết session vừa chết thuộc về ai / tier nào để ghi được dòng audit
-- `expired`. Không có hai field này thì đường đời phổ biến NHẤT (hết hạn tự
-- nhiên) không để lại sự kiện kết thúc nào trong `sessions_audit`, và câu hỏi
-- forensic mà B8 sinh ra để trả lời không trả lời được cho đa số phiên.
-- Cùng lý do với việc `session:{id}:pod` sống lâu hơn hash.
w('HSET', podPrefix .. pod,
  'state', 'claimed',
  'sessionId', ARGV[1],
  'userId', ARGV[2],
  'tier', ARGV[4],
  'updatedAt', ARGV[5])

-- Field khớp rediskeys.SessionFields — orchestrator ghi, gateway đọc cho authz
-- vế g. Ghi lệch tên field ở đây là authz so sánh với chuỗi rỗng trong im lặng.
w('HSET', KEYS[3],
  'userId', ARGV[2],
  'podName', pod,
  'namespace', ARGV[3],
  'status', 'CLAIMED',
  'tier', ARGV[4],
  'createdAt', ARGV[5],
  'expiresAt', ARGV[6],
  'revision', 1,
  'lastActiveAt', ARGV[5])

w('SET', KEYS[4], pod)
w('EXPIRE', KEYS[3], ARGV[7])
-- TTL DÀI HƠN hash có chủ ý: reaper tầng 1 nghe `__keyevent@0__:expired` của
-- `session:{id}`, và lúc event tới thì hash ĐÃ biến mất — không còn chỗ nào đọc
-- được podName để xoá pod. Con trỏ này sống thêm một khoảng grace chính là để
-- trả lời "session vừa hết hạn đang ở pod nào". Đặt bằng nhau là reaper mù.
w('EXPIRE', KEYS[4], ARGV[9])

if failed then
  undo()
  return redis.error_reply('claim: abort roi hoan tac: ' .. failed)
end

return pod

File: services/orchestrator/internal/pool/claim.go
// Package pool giữ warm-pool của sandbox pod (phase-1 1.B).
//
// File này là sản phẩm của spike 1.A-2: chứng minh claim atomic dưới đua
// 200 goroutine trên Redis THẬT. Warm-pool manager (B2) sẽ xây quanh nó.
package pool

import (
	"context"
	_ "embed"
	"errors"
	"fmt"
	"strconv"
	"time"

	"github.com/redis/go-redis/v9"

	"github.com/Nghaiz/DevOps_Learning_Platform/services/shared/rediskeys"
)

//go:embed claim.lua
var claimLua string

// claimScript được nạp một lần cho cả process. redis.NewScript tự làm đúng
// nghi thức EVALSHA-trước-EVAL-sau: Run() thử EVALSHA bằng SHA1 tính sẵn,
// gặp NOSCRIPT (Redis restart làm trống script cache) thì tự fallback EVAL —
// tức là nạp lại script trong cùng một lời gọi, không mất claim nào.
// Test TestClaimSurvivesScriptFlush chứng minh hành vi này trên Redis thật.
var claimScript = redis.NewScript(claimLua)

// ErrPoolEmpty là sentinel cho cold-path: pool rỗng KHÔNG phải lỗi hệ thống.
// Caller (B3) bắt nó để tự tạo pod đồng bộ thay vì claim từ pool.
var ErrPoolEmpty = errors.New("pool: không còn pod ấm trong pool:free")

// Trần của TTL. Không phải con số cho đẹp: `EXPIRE` từ chối giá trị vượt
// khoảng nó chấp nhận, và lỗi đó nổ ra ở lệnh CUỐI của claim.lua — sau khi
// session đã ghi xong. Script tự hoàn tác được, nhưng chặn từ biên thì rẻ hơn
// và cho thông báo hiểu được thay vì "invalid expire time in 'expire' command".
//
// 24h là trần kỹ thuật, KHÔNG phải chính sách: HARD_CAP thật (D11: 2h) do B5
// áp. Cái này chỉ chặn giá trị vô nghĩa lọt xuống Redis.
const maxTTLSeconds = int64(24 * 60 * 60)

// podPointerGrace là khoảng `session:{id}:pod` sống LÂU HƠN hash session.
//
// Reaper tầng 1 nghe `__keyevent@0__:expired` của `session:{id}`; lúc event
// tới thì hash ĐÃ biến mất và không còn chỗ nào đọc được podName để xoá pod.
// Con trỏ sống thêm khoảng này chính là thứ trả lời "session vừa hết hạn đang
// ở pod nào". Đặt hai TTL bằng nhau là reaper mù.
const podPointerGrace = 10 * time.Minute

// ClaimParams là đầu vào cho một lượt claim. Mọi field bắt buộc.
type ClaimParams struct {
	SessionID string
	UserID    string
	Namespace string
	Tier      string
	// NowUnix ghi vào createdAt/lastActiveAt/pod.updatedAt — truyền từ ngoài
	// để test kiểm được giá trị và để không gọi time.Now() trong đường nóng.
	NowUnix int64
	// ExpiresAtUnix ghi vào field expiresAt của hash session.
	ExpiresAtUnix int64
	// TTLSeconds đặt EXPIRE cho session:{id}.
	TTLSeconds int64
}

func (p ClaimParams) validate() error {
	// SessionID/UserID: định danh, qua cổng pattern của rediskeys.
	if err := rediskeys.ValidateID(p.SessionID); err != nil {
		return fmt.Errorf("pool: session id: %w", err)
	}
	// UserID không nằm trong key nào của script này, nhưng nó được ghi vào hash
	// mà gateway so sánh NGUYÊN VĂN cho authz vế g — cùng miền định danh với
	// Idem(userID, …), nên phải qua cùng một cổng.
	if err := rediskeys.ValidateID(p.UserID); err != nil {
		return fmt.Errorf("pool: user id: %w", err)
	}
	// Namespace đi ra khỏi Redis rồi thành một đoạn trong URL `pods/exec` mà
	// gateway gọi (G4 đọc field này thay vì hardcode). Một giá trị không kiểm
	// vượt biên tin cậy Redis → K8s API, nên nó qua cùng cổng với định danh.
	if err := rediskeys.ValidateID(p.Namespace); err != nil {
		return fmt.Errorf("pool: namespace: %w", err)
	}
	// Tier là enum của proto. Chuỗi tự do ở đây nghĩa là gõ sai tồn tại im lặng
	// trong Redis rồi đi thẳng ra FE.
	if !validTiers[p.Tier] {
		return fmt.Errorf("pool: tier %q không hợp lệ (cần một trong %v)", p.Tier, tierNames())
	}
	if p.NowUnix <= 0 {
		return fmt.Errorf("pool: NowUnix phải > 0 (nhận %d)", p.NowUnix)
	}
	if p.TTLSeconds <= 0 || p.TTLSeconds > maxTTLSeconds {
		return fmt.Errorf("pool: TTLSeconds phải trong (0, %d] (nhận %d)", maxTTLSeconds, p.TTLSeconds)
	}
	// Không kiểm thì tạo được session sinh ra đã hết hạn — gateway sẽ từ chối
	// mọi kết nối vào nó với 409 và không ai hiểu vì sao.
	if p.ExpiresAtUnix <= p.NowUnix {
		return fmt.Errorf("pool: ExpiresAtUnix (%d) phải sau NowUnix (%d)", p.ExpiresAtUnix, p.NowUnix)
	}
	return nil
}

// validTiers khớp enum SandboxTier của proto. UNSPECIFIED bị loại có chủ ý:
// comment trong session.proto yêu cầu fail-closed với giá trị đó.
var validTiers = map[string]bool{
	"SANDBOX_TIER_SYSBOX": true,
	"SANDBOX_TIER_KATA":   true,
}

func tierNames() []string {
	names := make([]string, 0, len(validTiers))
	for k := range validTiers {
		names = append(names, k)
	}
	return names
}

// Claim rút pod cũ nhất ĐANG FREE khỏi pool:free và gắn trọn bộ state cho
// session trong MỘT lượt Redis. Pool rỗng → ErrPoolEmpty.
//
// Script tự hoàn tác nếu một lệnh ghi lỗi giữa chừng, nên một lỗi trả về từ
// đây nghĩa là Redis KHÔNG còn state dở của lượt claim này — pod đã về lại
// pool:free và retry là an toàn.
func Claim(ctx context.Context, rdb redis.Scripter, p ClaimParams) (podName string, err error) {
	if err := p.validate(); err != nil {
		return "", err
	}
	sessionKey, err := rediskeys.Session(p.SessionID)
	if err != nil {
		return "", err
	}
	sessionPodKey, err := rediskeys.SessionPod(p.SessionID)
	if err != nil {
		return "", err
	}

	keys := []string{
		rediskeys.PoolFree,
		rediskeys.PoolClaimed,
		sessionKey,
		sessionPodKey,
		rediskeys.PoolQuarantine,
	}
	argv := []interface{}{
		p.SessionID,
		p.UserID,
		p.Namespace,
		p.Tier,
		strconv.FormatInt(p.NowUnix, 10),
		strconv.FormatInt(p.ExpiresAtUnix, 10),
		strconv.FormatInt(p.TTLSeconds, 10),
		rediskeys.PodPrefix,
		strconv.FormatInt(p.TTLSeconds+int64(podPointerGrace.Seconds()), 10),
	}

	pod, err := claimScript.Run(ctx, rdb, keys, argv...).Text()
	if errors.Is(err, redis.Nil) {
		// Lua `return nil` → redis trả bulk nil → go-redis map thành redis.Nil.
		return "", ErrPoolEmpty
	}
	if err != nil {
		return "", fmt.Errorf("pool: chạy claim.lua: %w", err)
	}
	return pod, nil
}
